engine/src/main/java/com/google/android/fhir/sync/upload/patch/StronglyConnectedPatches.kt

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.sync.upload.patch
+
+import kotlin.math.min
+
+internal object StronglyConnectedPatches {
+
+  /**
+   * Takes a [directedGraph] and computes all the strongly connected components in the graph.
+   *
+   * @return An ordered List of strongly connected components of the [directedGraph]. The SCCs are
+   *   topologically ordered which may change based on the ordering algorithm and the [Node]s inside
+   *   a SSC may be ordered randomly depending on the path taken by algorithm to discover the nodes.
+   */
+  fun scc(directedGraph: Graph): List<List<Node>> {
+    return findSCCWithTarjan(directedGraph)
+  }
+
+  /**
+   * Finds strongly connected components in topological order. See
+   * https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm.
+   */
+  private fun findSCCWithTarjan(diGraph: Graph): List<List<Node>> {
+    val nodeCount = (diGraph.keys + diGraph.values.flatten().toSet()).size
```

Comment:
Should'nt this be?
```suggestion
    val nodeCount = (diGraph.keys + diGraph.values.flatten()).toSet().size
```

Or actually the diGraph.keys will have all the keys already. Look at `createAdjacencyListForCreateReferences` where for each patchMapping we add a List<Node> which could also be empty.

Reply:
`diGraph.keys` is already a set. Its just to make sure that if ever the code is called with a graph with just the edge added, it doesn't fail.